lib/bigstringaf_stubs.c

```diff
@@ -83,3 +86,117 @@ bigstringaf_memcmp_string(value vba, value vba_off, value vstr, value vstr_off,
     int result = memcmp(buf1, buf2, len);
     return Val_int(result);
 }
+
+CAMLprim value
+bigstringaf_read(value vfd, value vba, value voff, value vlen)
+{
+    void *iobuf = ((char *)Caml_ba_data_val(vba)) + Unsigned_long_val(voff);
+#ifdef Handle_val
+    unsigned len;
+    int err;
+
+    if (Descr_kind_val(fd) == KIND_SOCKET) {
+      SOCKET s = Socket_val(fd);
+      caml_release_runtime_system();
+      if ((err = recv(s, iobuf, Unsigned_int_val(vlen), 0)) < 0)
+        err = WSAGetLastError();
+      else {
+        len = err;
+        err = 0;
+      }
+      caml_acquire_runtime_system();
+    } else {
+      HANDLE h = Handle_val(fd);
+      caml_release_runtime_system();
+      if (ReadFile(h, iobuf, Unsigned_int_val(vlen), &len, NULL))
+        err = 0;
+      else {
+check_error:
+        switch (err = GetLastError()) {
+          case ERROR_BROKEN_PIPE:
+            /* This is no error, but just a closed pipe. */
+            err = len = 0;
+            break;
+          case ERROR_MORE_DATA:
+#if 0
+            do {
+              char buf[1024];
+              unsigned dummy_len;
+              if (ReadFile(h, buf, sizeof(buf), &dummy_len, NULL))
+                break;
+              else
+                goto check_error;
+            } while (0);
+#else
+            err = 0;
+#endif
+          default:
+            break;
+        }
+      caml_acquire_runtime_system();
+    }
+    /* GetLastError() and WSAGetLastError() error numbers _are_ compatible,
+     * although not documented this behaviour will hopefully never change. */
+    if (err) {
+      win32_maperr(err);
+      uerror("read", Nothing);
+    }
+    else
+      return Val_int(len);
+
+#else
+    ssize_t ret;
+
+    caml_release_runtime_system();
```

Comment:
Releasing the lock should be optional, specified by an argument, in consideration for folks that want to do non-blocking calls.

Reply:
I suspect you got something wrong here. ``caml_release_runtime_system()`` won't make the call non-blocking. The syscalls will still be blocking, but _other_ OCaml threads will be allowed to run. If I'm mistaken here would you please explain what is there to gain in keeping the runtime lock?

Reply:
If the fd is set to nonblocking, then releasing the lock is more trouble than it's worth. Rather than checking the state of the fd, you can just have the caller tell the function what locking behavior is desired.